feat: implement runtime configuration package with multi-org support

dannykopping opened the pull request:
[RFC](https://www.notion.so/coderhq/Runtime-Configuration-d257df712dad421da9abb8d7ae8bd782?pvs=4).

# What this does

Adds a new `runtimeconfig` package for managing all runtime related configuration. The goal of this package is to create a consistent method for CRUD on runtime configuration (as opposed to startup values). 

The motivation and driver is mutli-org. Multi org settings must be configured at runtime. Instead of creating ad-hoc CRUD operations to a key/val store, this package creates a single code path for all settings.

# Design

The design uses a `Manager` singleton that produces `Resolvers` to enable a cache layer in the future. A pure singleton cannot be used as the `Resolver` requires a `database.Store`, which could be a `tx`. So the user of the package must create the `Resolver` on demand with their local instance of the `db`. 

## Usage

The design is intentionally created so that we can deprecate startup values in favor of runtime in the future. The `DeploymentEntry` type specifically uses startup values as a fallback.

This will allow us to move some startup values into runtime, which has been a sticky point for configuring aspects of coder for awhile now.

# Future work

### Caching layer :: https://github.com/coder/coder/issues/14586

A follow up to this PR will be a caching layer that uses pubsub to update it's cached values. Adding a cache layer will make all settings essentially free to read, allowing use in hot paths such as `httpmw` (as we currently do for startup values).

### RBAC layer :: https://github.com/coder/internal/issues/154

When caching is implemented, `dbauthz` will not be called for cache hits. So this layer, or some layer above, needs to be RBAC aware.

Maybe in the `initialize`? the `New`? The field could know it's own RBAC properties??


### Remove old runtime settings in favor of this

Pretty much all these sql queries: https://github.com/coder/coder/blob/e461d2773b5a2f1bd04ff899a922032b373e187f/coderd/database/queries/siteconfig.sql

Upside is we will remove a lot of database.Store queries (always nice to trim)

dannykopping committed:
Implementation

Signed-off-by: Danny Kopping <danny@coder.com>

Emyrk commented on coderd/runtimeconfig/config.go:
```diff
@@ -0,0 +1,97 @@
+}
+
+func (o *Entry[T]) Init(key string) {
+	o.val = create[T]()
```
Is this to prevent a nil derefence? Why do we need to `init` the value?

dannykopping replied:
I've adjusted the implementation to handle zero values more elegantly.

Emyrk commented on coderd/runtimeconfig/config.go:
```diff
@@ -0,0 +1,97 @@
+	return o.val
+}
+
+func (o *Entry[T]) Resolve(ctx context.Context, r Resolver) (T, error) {
+	return o.resolve(ctx, r)
+}
+
+func (o *Entry[T]) resolve(ctx context.Context, r Resolver) (T, error) {
```
Why have this method twice? Why not just `Resolve`?

dannykopping replied:
`Coalesce` calls it as well.

Emyrk commented on coderd/runtimeconfig/config.go:
```diff
@@ -0,0 +1,97 @@
+		return zero, err
+	}
+
+	inst := create[T]()
+	if err = inst.Set(val); err != nil {
+		return zero, xerrors.Errorf("instantiate new %T: %w", inst, err)
+	}
```
Is the purpose of this to always return a copy rather than the original?

Emyrk commented on coderd/runtimeconfig/config_test.go:
```diff
@@ -0,0 +1,123 @@
+		ctx := testutil.Context(t, testutil.WaitShort)
+		store := runtimeconfig.NewInMemoryStore()
+		resolver := runtimeconfig.NewOrgResolver(altOrg.ID, runtimeconfig.NewStoreResolver(store))
+		mutator := runtimeconfig.NewOrgMutator(altOrg.ID, runtimeconfig.NewStoreMutator(store))
```
Just another option, I haven't put much thought, so feel free to ignore.

Another syntax could be `runtimeconfig.NewStoreMutator(store).WithOrg(orgID)`.

Just using a chain syntax rather than a new `NewOrgMutator` function :man_shrugging:. I just don't imagine we will have use anything more than the StoreMutator.

Emyrk commented on coderd/runtimeconfig/spec.go:
```diff
@@ -0,0 +1,25 @@
+
+import "context"
+
+type Initializer interface {
+	Init(key string)
+}
```
What is this used for?

Emyrk commented on coderd/runtimeconfig/store.go:
```diff
@@ -0,0 +1,53 @@
+	DeleteRuntimeSetting(ctx context.Context, key string) error
+}
+
+type InMemoryStore struct {
+	mu    sync.Mutex
+	store map[string]string
+}
```
If we just embed this in the dbmem, that is pretty nice :+1:

Emyrk reviewed (commented):
Is the look aside cache just going to be a wrapper on the resolver?

I like the interface simplicity. The only thing that I don't 100% follow is the usage of reflect. It would be good to drop comments on why it is required.

dannykopping committed:
Support zero values

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Add database queries against site_configs

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Document usage in test

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Touchups

Signed-off-by: Danny Kopping <danny@coder.com>

Emyrk committed:
make gen to generate dbmock functions

dannykopping committed:
make lint/fmt

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Initializer interface, rename "key" to "name"

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Manager interface

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Manager, with scoping

Signed-off-by: Danny Kopping <danny@coder.com>

Emyrk committed:
linting errors

Emyrk committed:
chore: add runtimeconfig manager to options for plumbing (#14562)

Emyrk committed:
linting, remove unused ctx

Emyrk commented on cli/server.go:
```diff
@@ -820,6 +821,14 @@ func (r *RootCmd) Server(newAPI func(context.Context, *coderd.Options) (*coderd.
+			// Note: This happens before dbauthz, which is really unfortunate.
+			// dbauthz is configured in `Coderd.New()`, but we need the manager
+			// at this level for notifications. We might have to move some init
+			// code around.
```
@spikecurtis yes, I fixed this :+1:

Emyrk committed:
chore: refactor entry into deployment and runtime (#14575)

* chore: refactor entry into deployment and runtime

RuntimeEntry has no startup value, and omits functions required
to be serpent compatible.

Co-authored-by: Danny Kopping <danny@coder.com>

---------

Co-authored-by: Danny Kopping <danny@coder.com>

Emyrk committed:
chore: comments, cleanup, error fixes

johnstcn commented on coderd/database/dbauthz/dbauthz.go:
```diff
@@ -3906,6 +3920,13 @@ func (q *querier) UpsertProvisionerDaemon(ctx context.Context, arg database.Upse
 }
 
+func (q *querier) UpsertRuntimeConfig(ctx context.Context, arg database.UpsertRuntimeConfigParams) error {
+	if err := q.authorizeContext(ctx, policy.ActionCreate, rbac.ResourceSystem); err != nil {
```
Does the usage of `rbac.ResourceSystem` here mean that only site-wide owners can edit runtime config?

Emyrk replied:
I intentionally am making this vague atm. 

Once we start adding APIs to interact with the configs, we will probably have to write custom authorize calls depending on the setting. 

The goal of this layer is to be able to add a consistent caching layer infront, so we don't need to implement some pubsub notify go routine for each setting. When we implement the caching layer, it will skip calling the `dbauthz` layer entirely on cache hits.

So the RBAC story has to be solved for this layer, and I'm kicking the can on it for now. This is the core, basic CRUD.

Emyrk replied:
Here is an example where we skip crud on cached values today:

https://github.com/coder/coder/blob/e461d2773b5a2f1bd04ff899a922032b373e187f/agent/agent.go#L713-L740

So this PR does not address it, but it is not addressed in current day runtime configs

dannykopping replied:
Can we add a TODO/FIXME here, with an issue link?

Emyrk replied:
I don't think we need to since it requires system auth right now, which is the highest.

So any user of this package essentially needs to solve it on their own until a general solution is created. (I might whip up something).

So this defaults to "secure" essentially.

johnstcn replied:
If we're kicking the work down the road, it's definitely good idea to create an issue to track it. Doesn't have to be in the `coder/coder` repo.

Emyrk replied:
Added issue: https://github.com/coder/coder/issues/14587

Emyrk committed:
increase test coverage:

Emyrk committed:
linting complained about an unhandled Println error

dannykopping commented on coderd/database/dbmem/dbmem.go:
```diff
@@ -3505,6 +3515,18 @@ func (q *FakeQuerier) GetReplicasUpdatedAfter(_ context.Context, updatedAt time.
 }
 
+func (q *FakeQuerier) GetRuntimeConfig(_ context.Context, key string) (string, error) {
+	q.mutex.Lock()
```
Nit: we could use `RLock/RUnlock` here.

Emyrk replied:
The fake db uses `Lock` to find any racy stuff with incorrect db usage. Like using a `db` inside a `tx`.

We essentially only support a single "connection" to the db at any given time for `dbmem`.

dannykopping commented:
@Emyrk good to go :+1:

Emyrk commented on coderd/runtimeconfig/manager.go:
```diff
@@ -0,0 +1,28 @@
+// TODO: Implement caching layer.
+type StoreManager struct{}
+
+func NewStoreManager() Manager {
```
Called it `Manager` and dropped the interface

Emyrk commented on coderd/runtimeconfig/spec.go:
```diff
@@ -0,0 +1,47 @@
+
+type Resolver interface {
+	// GetRuntimeSetting gets a runtime setting by name.
+	GetRuntimeSetting(ctx context.Context, name string) (string, error)
+	// UpsertRuntimeSetting upserts a runtime setting by name.
+	UpsertRuntimeSetting(ctx context.Context, name, val string) error
+	// DeleteRuntimeSetting deletes a runtime setting by name.
+	DeleteRuntimeSetting(ctx context.Context, name string) error
```
To make the db query name? Makes sense to me. We don't use the word `Setting` anywhere else.

Emyrk commented on coderd/runtimeconfig/resolver.go:
```diff
@@ -0,0 +1,92 @@
+	return NamespacedResolver{ns: orgID.String(), wrapped: wrapped}
+}
+
+func (m NamespacedResolver) GetRuntimeSetting(ctx context.Context, key string) (string, error) {
+	return m.wrapped.GetRuntimeSetting(ctx, m.namespacedKey(key))
+}
+
+func (m NamespacedResolver) UpsertRuntimeSetting(ctx context.Context, key, val string) error {
+	return m.wrapped.UpsertRuntimeSetting(ctx, m.namespacedKey(key), val)
+}
+
+func (m NamespacedResolver) DeleteRuntimeSetting(ctx context.Context, key string) error {
+	return m.wrapped.DeleteRuntimeSetting(ctx, m.namespacedKey(key))
+}
+
+func (m NamespacedResolver) namespacedKey(k string) string {
+	return fmt.Sprintf("%s:%s", m.ns, k)
+}
```
I intentionally did not because the `ns` should not be mutated in any of it's functions. `StoreResolver` does not use a pointer either.

Emyrk commented on coderd/runtimeconfig/entry.go:
```diff
@@ -0,0 +1,95 @@
+// UnsetRuntimeValue removes the runtime value from the store.
+func (e *RuntimeEntry[T]) UnsetRuntimeValue(ctx context.Context, m Resolver) error {
+	name, err := e.name()
+	if err != nil {
+		return err
+	}
```
Yea, let's do it. Will add :+1:

Emyrk committed:
drop manager interface, wrap errors

Emyrk committed:
rename NewStoreManager -> NewManager

johnstcn reviewed (approved):
Approving to unblock.

dannykopping commented:
https://github.com/coder/internal/issues/37#issuecomment-2352257029
